src/Sylius/Component/Product/Resolver/DefaultProductVariantResolver.php

```diff
@@ -13,17 +13,34 @@
 
 namespace Sylius\Component\Product\Resolver;
 
+use Sylius\Component\Product\Repository\ProductVariantRepositoryInterface;
 use Sylius\Component\Product\Model\ProductInterface;
 use Sylius\Component\Product\Model\ProductVariantInterface;
 
 final class DefaultProductVariantResolver implements ProductVariantResolverInterface
 {
+    public function __construct(private ?ProductVariantRepositoryInterface $productVariantRepository = null)
+    {
+    }
+
     public function getVariant(ProductInterface $subject): ?ProductVariantInterface
     {
-        if ($subject->getEnabledVariants()->isEmpty()) {
-            return null;
+        if (!$this->productVariantRepository) {
+            trigger_deprecation('sylius/product', '1.12', sprintf('Not passing an instance of %s as constructor argument for %s is deprecated as of Sylius 1.12 and will be mandatory in 1.13.', ProductVariantRepositoryInterface::class, self::class));
+
+            if ($subject->getEnabledVariants()->isEmpty()) {
+                return null;
+            }
+
+            return $subject->getEnabledVariants()->first();
         }
 
-        return $subject->getEnabledVariants()->first();
+        /** @var ProductVariantInterface|null $productVariant */
+        $productVariant = $this->productVariantRepository->findOneBy([
+            'product' => $subject->getId(),
```

Comment:
We should return early when a product has no id.

Reply:
I think we should leave this as it is.
You should only use resolver when ID is present, because we are using DB to fetch variant.
If id is optional, than it can lead to confusion.

Example if we do early return:
Product is created in memory and we set variants for his. Everything is still in memory, results aren't flushed to the database.
Method `$product->getVariants()` will return variants, but trying to resolve variants from the default resolver, will result in `null`.

But if we leave it like it is, resolver will throw an error, which is clear indication that something is wrong.

Reply:
Even when left as is we're still returning null when the product was not saved, the only difference with an early return would be not triggering doctrine. But now that I think about it, how about falling back to the previous logic in case the product doesn't have an id. The optimization would still be in effect, while the previous behaviour stays the same and we maintain one access point for getting a variant.

Reply:
Agree with you.